Race condition when starting container causes IndexOutOfBoundsException when binding ports

### Description
An intermittent issue can be seen when starting containers. It seems to be a race condition that is exacerbated when running maven with debug logging `-X`. Without the debug logging containers fail to start with  
```
[ERROR] DOCKER> Error occurred during container startup, shutting down...
```

The full stack trace of the exception is:
```
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
    at jdk.internal.util.Preconditions.outOfBounds (Preconditions.java:64)
    at jdk.internal.util.Preconditions.outOfBoundsCheckIndex (Preconditions.java:70)
    at jdk.internal.util.Preconditions.checkIndex (Preconditions.java:248)
    at java.util.Objects.checkIndex (Objects.java:372)
    at java.util.ArrayList.get (ArrayList.java:459)
    at com.google.gson.JsonArray.get (JsonArray.java:194)
    at io.fabric8.maven.docker.model.ContainerDetails.createPortBindings (ContainerDetails.java:200)
    at io.fabric8.maven.docker.model.ContainerDetails.getPortBindings (ContainerDetails.java:139)
    at io.fabric8.maven.docker.service.RunService.updateMappedPortsAndAddresses (RunService.java:461)
    at io.fabric8.maven.docker.service.RunService.createAndStartContainer (RunService.java:156)
    at io.fabric8.maven.docker.service.helper.StartContainerExecutor.startContainer (StartContainerExecutor.java:47)
    at io.fabric8.maven.docker.StartMojo.lambda$startImage$0 (StartMojo.java:299)
```

When adding some logging to the maven plugin I can see that during the the `io.fabric8.maven.docker.model.ContainerDetails.createPortBindings` method call the ports will usually have some mappings like this:
```
Ports:{"5432/tcp":[{"HostIp":"0.0.0.0","HostPort":"54290"}],}
```
However sometimes the array of mapped ports will be empty, e.g.
```
Ports:{"5432/tcp":[]}
```

This causes the exception as the array is expected to not be empty and the first element retrieved.

I believe the call to the api in `io.fabric8.maven.docker.service.RunService.updateMappedPortsAndAddresses` is sometimes returning this empty data for the port if called too early.

The solution I would propose would be to retry the API call and the `createPortBindings` activity if an empty array is encountered.

### Info

* docker-maven-plugin version : 0.34.1
* Maven version (`mvn -v`) : 
```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/Cellar/maven/3.6.3_1/libexec
Java version: 11.0.10, vendor: AdoptOpenJDK, runtime: /Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "10.14.6", arch: "x86_64", family: "mac"
```
* Docker version : 3.2.2
* If it's a bug, how to reproduce : The issue is intermittent so impossible to consistently reproduce. It appeared to be more easily reproducible when running `mvn verify` with `-X` flag.

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -13,6 +13,7 @@
   - Add a skipPom parameter, skipping a project if packaging is pom ([1388](https://github.com/fabric8io/docker-maven-plugin/pull/1388))
   - Add support for config to specify isolation technology for container ([1376](https://github.com/fabric8io/docker-maven-plugin/pull/1376))
   - Add Docker build cache friendly example utilizing Spring Boot Layered JAR and Maven Assembly Plugin ([1412](https://github.com/fabric8io/docker-maven-plugin/pull/1412))
+  - Retry port mapping to avoid race condition where complete port information may not be initially available in Docker Engine 20.10.5 ([1447](https://github.com/fabric8io/docker-maven-plugin/pull/1447))
   
 * **0.34.1** (2020-09-27)
   - Fix NPE with "skipPush" and no build configuration given ([#1381](https://github.com/fabric8io/docker-maven-plugin/issues/1381))

--- a/pom.xml
+++ b/pom.xml
@@ -183,6 +183,12 @@
       <artifactId>plexus-interpolation</artifactId>
       <version>1.24</version>
     </dependency>
+    
+    <dependency>
+      <groupId>net.jodah</groupId>
+      <artifactId>failsafe</artifactId>
+      <version>2.4.0</version>
+    </dependency>
 
 
     <!-- =============================================================================== -->

--- a/src/main/java/io/fabric8/maven/docker/model/ContainerDetails.java
+++ b/src/main/java/io/fabric8/maven/docker/model/ContainerDetails.java
@@ -8,6 +8,8 @@ import java.util.Map;
 import java.util.Set;
 
 import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonArray;
 import com.google.gson.JsonObject;
 
 
@@ -196,8 +198,12 @@ public class ContainerDetails implements Container {
             if (ports.get(port).isJsonNull()) {
                 addPortMapping(port, (PortBinding) null, portBindings);
             } else {
+                JsonArray hostMappings = ports.getAsJsonArray(port);
+                if (hostMappings.isJsonNull() || hostMappings.size() == 0) {
+                    throw new PortBindingException(port, ports);
+                }
                 // use the first entry in the array
-                JsonObject hostConfig = ports.getAsJsonArray(port).get(0).getAsJsonObject();
+                JsonObject hostConfig = hostMappings.get(0).getAsJsonObject();
                 addPortMapping(port, hostConfig, portBindings);
             }
         }

--- a/src/main/java/io/fabric8/maven/docker/model/PortBindingException.java
+++ b/src/main/java/io/fabric8/maven/docker/model/PortBindingException.java
@@ -1,0 +1,9 @@
+package io.fabric8.maven.docker.model;
+
+import com.google.gson.JsonObject;
+
+public class PortBindingException extends RuntimeException {
+    public PortBindingException(String port, JsonObject portDetails) {
+        super(String.format("Failed to create binding for port '%s'. Container ports: %s", port, portDetails));
+    }
+}

--- a/src/main/java/io/fabric8/maven/docker/service/RunService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/RunService.java
@@ -20,6 +20,7 @@ import static io.fabric8.maven.docker.util.VolumeBindingUtil.resolveRelativeVolu
  */
 
 import java.io.File;
+import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -55,6 +56,7 @@ import io.fabric8.maven.docker.model.Container;
 import io.fabric8.maven.docker.model.ContainerDetails;
 import io.fabric8.maven.docker.model.ExecDetails;
 import io.fabric8.maven.docker.model.Network;
+import io.fabric8.maven.docker.model.PortBindingException;
 import io.fabric8.maven.docker.util.ContainerNamingUtil;
 import io.fabric8.maven.docker.util.EnvUtil;
 import io.fabric8.maven.docker.util.GavLabel;
@@ -62,6 +64,8 @@ import io.fabric8.maven.docker.util.Logger;
 import io.fabric8.maven.docker.util.StartOrderResolver;
 import io.fabric8.maven.docker.wait.WaitTimeoutException;
 import io.fabric8.maven.docker.wait.WaitUtil;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.RetryPolicy;
 
 
 /**
@@ -464,12 +468,23 @@ public class RunService {
     }
 
     private void updateMappedPortsAndAddresses(String containerId, PortMapping mappedPorts) throws DockerAccessException {
-        Container container = queryService.getMandatoryContainer(containerId);
-        if (container.isRunning()) {
-            mappedPorts.updateProperties(container.getPortBindings());
-        } else {
-            log.warn("Container %s is not running anymore, can not extract dynamic ports",containerId);
-        }
+        RetryPolicy<Void> retryPolicy = new RetryPolicy<Void>()
+                .withMaxAttempts(10)
+                .withDelay(Duration.ofMillis(250))
+                .handle(PortBindingException.class)
+                .onFailedAttempt(f -> log.debug("Failed to update mapped ports for container %s (attempt %d), retrying", 
+                        containerId, f.getAttemptCount()))
+                .onRetriesExceeded(f -> log.warn("Failed to update mapped ports for container %s after %d retries", 
+                        containerId, f.getAttemptCount()));
+        
+        Failsafe.with(retryPolicy).run(() -> {
+            Container container = queryService.getMandatoryContainer(containerId);
+            if (container.isRunning()) {
+                mappedPorts.updateProperties(container.getPortBindings());
+            } else {
+                log.warn("Container %s is not running anymore, can not extract dynamic ports", containerId);
+            }
+        });
     }
 
     private void shutdown(ContainerTracker.ContainerShutdownDescriptor descriptor, boolean keepContainer, boolean removeVolumes)

--- a/src/main/java/io/fabric8/maven/docker/service/RunService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/RunService.java
@@ -21,6 +21,7 @@ import static io.fabric8.maven.docker.util.VolumeBindingUtil.resolveRelativeVolu
 
 import java.io.File;
 import java.time.Duration;
+import java.time.temporal.ChronoUnit;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -469,8 +470,8 @@ public class RunService {
 
     private void updateMappedPortsAndAddresses(String containerId, PortMapping mappedPorts) throws DockerAccessException {
         RetryPolicy<Void> retryPolicy = new RetryPolicy<Void>()
-                .withMaxAttempts(10)
-                .withDelay(Duration.ofMillis(250))
+                .withMaxAttempts(20)
+                .withBackoff(10, 100, ChronoUnit.MILLIS)
                 .handle(PortBindingException.class)
                 .onFailedAttempt(f -> log.debug("Failed to update mapped ports for container %s (attempt %d), retrying", 
                         containerId, f.getAttemptCount()))

--- a/src/test/java/io/fabric8/maven/docker/model/ContainerDetailsTest.java
+++ b/src/test/java/io/fabric8/maven/docker/model/ContainerDetailsTest.java
@@ -133,6 +133,13 @@ public class ContainerDetailsTest {
         thenValidateContainer();
     }
 
+    @Test(expected = PortBindingException.class)
+    public void testCreateContainerWithEmptyPortBindings() throws Exception {
+        givenAContainerWithUnboundPorts();
+        whenCreateContainer();
+        container.getPortBindings();
+    }
+
     private JsonArray createHostIpAndPort(int port, String ip) {
         JsonObject object = new JsonObject();
 
@@ -171,6 +178,15 @@ public class ContainerDetailsTest {
 
     private void givenAContainerWithoutPorts() {
         json.add(ContainerDetails.NETWORK_SETTINGS, new JsonObject());
+    }
+    
+    private void givenAContainerWithUnboundPorts() {
+        JsonObject ports = new JsonObject();
+        ports.add("80/tcp", new JsonArray());
+        ports.add("52/udp", new JsonArray());
+        JsonObject networkSettings = new JsonObject();
+        networkSettings.add(ContainerDetails.PORTS, ports);
+        json.add(ContainerDetails.NETWORK_SETTINGS, networkSettings);
     }
 
     private void givenContainerData() {

--- a/src/test/java/io/fabric8/maven/docker/service/RunServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/RunServiceTest.java
@@ -371,7 +371,7 @@ public class RunServiceTest {
         new Expectations() {{
             docker.createContainer(withAny(new ContainerCreateConfig("img")), anyString); result = "containerId";
             portMapping.needsPropertiesUpdate(); result = true;
-            queryService.getMandatoryContainer("containerId"); result = container; times = 10;
+            queryService.getMandatoryContainer("containerId"); result = container; times = 20;
             container.isRunning(); result = true;
             container.getPortBindings(); result = new PortBindingException("5432/tcp", new Gson().fromJson("{\"5432/tcp\": []}", JsonObject.class));
         }};

--- a/src/test/java/io/fabric8/maven/docker/service/RunServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/RunServiceTest.java
@@ -1,10 +1,14 @@
 package io.fabric8.maven.docker.service;
 
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
 import com.google.gson.JsonObject;
 
 import io.fabric8.maven.docker.access.ExecException;
 import io.fabric8.maven.docker.config.StopMode;
 import io.fabric8.maven.docker.config.VolumeConfiguration;
+import io.fabric8.maven.docker.model.Container;
+import io.fabric8.maven.docker.model.PortBindingException;
 import io.fabric8.maven.docker.util.GavLabel;
 import org.apache.commons.io.IOUtils;
 import org.apache.maven.execution.MavenSession;
@@ -18,6 +22,7 @@ import java.io.IOException;
 import java.lang.reflect.Field;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.Date;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -42,11 +47,11 @@ import io.fabric8.maven.docker.util.JsonFactory;
 import io.fabric8.maven.docker.util.Logger;
 import mockit.Expectations;
 import mockit.Mocked;
+import mockit.Verifications;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertThrows;
 import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
 /**
  * This test need to be refactored. In fact, testing Mojos must be setup correctly at all. Blame on me that there are so
@@ -335,6 +340,42 @@ public class RunServiceTest {
         long start = System.currentTimeMillis();
         runService.stopContainer(container, createImageConfigWithStopMode(StopMode.kill), false, false);
         assertTrue("No wait", System.currentTimeMillis() - start < SHUTDOWN_WAIT);
+    }
+    
+    @Test
+    public void retryIfInsufficientPortBindingInformation(
+            @Mocked Container container,
+            @Mocked ImageConfiguration imageConfiguration,
+            @Mocked PortMapping portMapping
+    ) throws DockerAccessException {
+        new Expectations() {{
+            docker.createContainer(withAny(new ContainerCreateConfig("img")), anyString); result = "containerId";
+            portMapping.needsPropertiesUpdate(); result = true;
+            queryService.getMandatoryContainer("containerId"); result = container; times = 2;
+            container.isRunning(); result = true;
+            container.getPortBindings(); result = new PortBindingException("5432/tcp", new Gson().fromJson("{\"5432/tcp\": []}", JsonObject.class));
+            returns(ImmutableMap.of("5432/tcp", new Container.PortBinding(56741, "0.0.0.0")), ImmutableMap.of("5432/tcp", new Container.PortBinding(56741, "0.0.0.0")));
+        }};
+        String containerId = runService.createAndStartContainer(imageConfiguration, portMapping, new GavLabel("Im:A:Test"), properties, getBaseDirectory(), "blah", new Date());
+        new Verifications() {{
+            assertEquals("containerId", containerId);
+        }};
+    }
+
+    @Test(expected = PortBindingException.class)
+    public void failAfterRetryingIfInsufficientPortBindingInformation(
+            @Mocked Container container,
+            @Mocked ImageConfiguration imageConfiguration,
+            @Mocked PortMapping portMapping
+    ) throws DockerAccessException {
+        new Expectations() {{
+            docker.createContainer(withAny(new ContainerCreateConfig("img")), anyString); result = "containerId";
+            portMapping.needsPropertiesUpdate(); result = true;
+            queryService.getMandatoryContainer("containerId"); result = container; times = 10;
+            container.isRunning(); result = true;
+            container.getPortBindings(); result = new PortBindingException("5432/tcp", new Gson().fromJson("{\"5432/tcp\": []}", JsonObject.class));
+        }};
+        runService.createAndStartContainer(imageConfiguration, portMapping, new GavLabel("Im:A:Test"), properties, getBaseDirectory(), "blah", new Date());
     }
 
     private ImageConfiguration createImageConfig(int wait, int kill) {

